### PR TITLE
Adds new PTO backoff schedule

### PR DESCRIPTION
**Pull request by janaiyengar**
This PR uses PTOs to speculatively retransmit, to minimize latency. The number of speculative PTOs is configurable by the application via num_speculative_ptos. This PR does two things:

(i) At a new tail of a transfer, the sender retransmits speculatively, at PTO^x intervals, where x starts at -num_speculative_ptos, increasing upto 1 . This enables early recovery from potential losses at the tail at the cost of some redundant transmissions. So, at a tail, with num_speculative_ptos set to 2, the backoff schedule looks like this: 0.25, 0.5, 1, 2, 4, 8, ...

(ii) When not at a new tail, the sender sets the PTO alarm without any additional backoff for x intervals. This enables quick recovery from short blackouts before starting to exponentially back off. So, mid-transfer, with num_speculative_ptos set to 2, the backoff schedule looks like this: 1, 1, 1, 2, 4, 8, ...

A new tail is defined as follows:
- sender is not in PTO recovery, 
- there is no stream data to send, and
- new application data was sent since the last tail.

The number of packets sent during the speculative PTO events is reduced to 1.  Once the exponential backoff schedule is at 1 PTO, number of packets sent becomes 2, to compensate for the increased cost of probe loss and subsequent exponential backoff.

A new performant profile is added, that sets num_speculative_ptos to 2.

This PR also includes some other cleanup.

### CODE REVIEW

File: lib/quicly.c [hunk 1]
```diff
@@ -900,15 +900,18 @@ static void update_idle_timeout(quicly_conn_t *conn, int is_in_receive)
...
 }
 
 static void update_loss_alarm(quicly_conn_t *conn)
 {
-    quicly_loss_update_alarm(&conn->egress.loss, now, conn->egress.last_retransmittable_sent_at,
-                             conn->egress.sentmap.bytes_in_flight != 0 || conn->super.peer.address_validation.send_probe);
+    quicly_loss_update_alarm(
+        &conn->egress.loss, now, conn->egress.last_retransmittable_sent_at,
+        conn->egress.sentmap.bytes_in_flight != 0 || conn->super.peer.address_validation.send_probe,
+        conn->super.ctx->stream_scheduler->can_send(conn->super.ctx->stream_scheduler, conn, /*including_new_data=*/0));
```
Comment (kazuho) [inline lib/quicly.c]: I think that `including_new_data` should be set to 1.

We want to arm the aggressive timer only when there's nothing to send. Either having new data or data to "re"transmit is not the case.

File: include/quicly/loss.h [hunk 2]
```diff
@@ -158,9 +159,9 @@ inline void quicly_rtt_update(quicly_rtt_t *rtt, uint32_t latest_rtt, uint32_t a
     assert(rtt->smoothed != 0);
 }
 
-inline uint32_t quicly_rtt_get_pto(quicly_rtt_t *rtt, uint32_t max_ack_delay)
+inline uint32_t quicly_rtt_get_pto(quicly_rtt_t *rtt, uint32_t max_ack_delay, uint32_t min_pto)
 {
-    return rtt->smoothed + (rtt->variance != 0 ? rtt->variance * 4 : 1) + max_ack_delay;
+    return rtt->smoothed + (rtt->variance != 0 ? rtt->variance * 4 : min_pto) + max_ack_delay;
```
Comment (kazuho) [inline include/quicly/loss.h]: Thank you for catching this.

`min_pto` has been a configuration variable, but we have hard-coded the value here.

File: include/quicly/loss.h [hunk 3]
```diff
@@ -235,10 +245,12 @@ inline int quicly_loss_on_alarm(quicly_loss_t *r, uint64_t largest_sent, uint64_
         *restrict_sending = 0;
         return quicly_loss_detect_loss(r, largest_acked, do_detect);
     }
-    /* PTO. Send at least and at most 2 packets. */
+    /* PTO. Send at least and at most 1 packet during early (and aggressive) probing and 2 packets otherwise. */
     ++r->pto_count;
```
Comment (kazuho) [inline include/quicly/loss.h]: Do we need to set `r->pto_count` to `QUICLY_AGRESSIVE_PROBE_THRESHOLD` if it was zero and if there is some pending data to be sent?

**Comment by janaiyengar** [comment]
@kazuho : I've modified according to our discussion and also done some cleanup. PTAL. (I'm yet to test it, so don't merge it yet.)

File: include/quicly/defaults.h [hunk 4]
```diff
@@ -28,7 +28,7 @@ extern "C" {
 
 #include "quicly.h"
 
-extern const quicly_context_t quicly_default_context;
+extern const quicly_context_t quicly_spec_context;
```
Comment (kazuho) [inline include/quicly/defaults.h]: I'd assume that you'd also want to declare `quicly_performant_context` here.

File: include/quicly.h [hunk 5]
```diff
@@ -149,9 +148,9 @@ typedef enum en_quicly_event_attribute_type_t {
     QUICLY_EVENT_ATTRIBUTE_END_OF_RECOVERY,
     QUICLY_EVENT_ATTRIBUTE_BYTES_IN_FLIGHT,
     QUICLY_EVENT_ATTRIBUTE_CWND,
+    QUICLY_EVENT_ATTRIBUTE_NUMPTO,
```
Comment (kazuho) [inline include/quicly.h]: maybe add an underscore and make it "NUM_PTO"?

File: include/quicly/loss.h (hunk 3 above)
Comment (kazuho) [reply include/quicly/loss.h]: My bad, the design always increments `pto_count` from one, but applies different scale factors based on if there's something to send.

**Comment by janaiyengar** [comment]
@kazuho: I've rewritten the PR as we discussed. PTAL.

File: include/quicly/loss.h [hunk 6]
```diff
@@ -44,11 +44,16 @@ typedef struct quicly_loss_conf_t {
      * The default RTT used before an RTT sample is taken.
      */
     uint32_t default_initial_rtt;
+    /**
+     * Number of speculative PTOs at the end of a window. This must not be set to more than 3.
+     */
+    uint32_t num_speculative_ptos;
```
Comment (kazuho) [inline include/quicly/loss.h]: Maybe `uint8_t`? Other than that, the PR looks awesome. This is superb!

Comment (janaiyengar) [reply include/quicly/loss.h]: Yeah, good point. Done.

**Comment by kazuho** [comment]
Merged!

**Pull request merged**